nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py

```diff
@@ -193,9 +201,36 @@ def loss_func(self, pi_logprobs, ref_logprobs, labels, average_log_probs=False):
         rewards = self.get_reduced_masked_logps(
             pi_logprobs - ref_logprobs, labels, average_log_probs=average_log_probs
         )
-        chosen_rewards, reject_rewards = self.split_output_tensor(self.ref_policy_kl_penalty * rewards)
 
-        loss = -torch.nn.functional.logsigmoid(chosen_rewards - reject_rewards)
+        chosen_rewards, reject_rewards = self.split_output_tensor(rewards)
+        logits = chosen_rewards - reject_rewards
+        assert self.preference_loss in ["dpo", "ipo", "kto"]
+        if self.preference_loss == "dpo":
+            loss = (
+                -torch.nn.functional.logsigmoid(self.ref_policy_kl_penalty * logits) * (1.0 - self.label_smoothing)
+                - torch.nn.functional.logsigmoid(-self.ref_policy_kl_penalty * logits) * self.label_smoothing
+            )
+        elif self.preference_loss == "ipo":
+            # self.ref_policy_kl_penalty is the tau parameter in the paper
+            loss = (logits - 1.0 / (2.0 * self.ref_policy_kl_penalty)) ** 2
+        elif self.preference_loss == "kto":
+            rewards_kl = self.get_reduced_masked_logps(pi_logprobs - ref_logprobs, labels, average_log_probs=True)
+            chosen_kl, reject_kl = self.split_output_tensor(rewards_kl)
+            loss = torch.cat(
```

Comment:
another nit: I'm not sure what the performance impact of this is but this is creating a tensor and then filling it in with these 2 things followed by a mean -> why not do .sum() on each of these tensors and then divide it ourselves?

I think it also makes it more readable that way

Reply:
Changed it!